tests/cpp/test_multidevice_matmul.cpp

```diff
@@ -404,4 +404,167 @@ TEST_F(DistributedMatmulTest, Matmul_LayoutNT_ReduceScatter) {
                                     ->heuristic();
   EXPECT_EQ(heuristic, ScheduleHeuristic::ExprEval);
 }
+
+TEST_F(DistributedMatmulTest, MLP_Layer) {
+  std::unique_ptr<Fusion> fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+  auto mesh = DeviceMesh::createForNumDevices(communicator_->size());
+
+  int64_t sb = 64; // sequence * batch
+  int64_t h = 128;
+  int64_t h4 = 4 * h;
+
+  // TODO: error with dynamic shape
+  // C++ exception with description "ext_opt.hasValue() INTERNAL ASSERT FAILED
+  // at "csrc/dynamic_transform.cpp":276, Could not evaluate dynamic extent: i3
+  // Exception raised from DynamicTransformConcretizationInfo at
+  // csrc/dynamic_transform.cpp:276
+  TensorView* x = makeContigConcreteTensor({sb, h}, DataType::BFloat16);
+  TensorView* w0 = makeContigConcreteTensor(
+      {num_devices_, h4 / num_devices_, h}, DataType::BFloat16);
+  TensorView* b0 = makeContigConcreteTensor(
+      {num_devices_, h4 / num_devices_}, DataType::BFloat16);
+  TensorView* w1 = makeContigConcreteTensor(
+      {num_devices_, h, h4 / num_devices_}, DataType::BFloat16);
+  TensorView* b1 = makeContigConcreteTensor({h}, DataType::BFloat16);
+  fusion->addInput(x);
+  fusion->addInput(w0);
+  fusion->addInput(b0);
+  fusion->addInput(w1);
+  fusion->addInput(b1);
+
+  // Linear #1
+  // Notes: Manually breaking down the linear layer into nvfuser primitives
+  TensorView* linear_int0 = broadcast(x, {true, false, true, false});
+  TensorView* linear_int1 = broadcast(w0, {false, true, false, false});
+  TensorView* linear_int2 = mul(linear_int0, linear_int1);
+  TensorView* linear_int3 = sum(linear_int2, {-1});
```

Comment:
Use matmul or linear? cc @Priya2698

Reply:
The first linear layer can be replaced with a linear op because there is no communication in between the bias add and local matmul. The second linear layer will have to be a matmul since the pattern is matmul + allreduce + bias add. 
I do have tests with replacing the second linear layer with matmul, but it was generating very high error. My assumption is that the `testValidate` isn't meant to be used when chaining multiple matmuls together? I also fiddled around with the datatypes of the matmul (i.e. bfloat, float, double) but the error was still high.

Reply:
> matmul + allreduce + bias add

Is it crazy to do `matmul + add half bias + all_reduce`? This may be slightly faster, but we can explore that later. 

> it was generating very high error

Does the second linear layer in the PR as is map to pointwise+reduction or aten matmul? If the former, we'll need to fix it soon because pointwise+reduction is likely slower. If the latter, I'll be curious why aten matmul via mul+sum is low error than aten matmul via matmul.

Reply:
> matmul + allreduce + bias add

Also, mainstream (e.g. llama) doesn't seem to use bias. So it might not be relevant. 

> Currently, the second linear layer uses nvfuser matmul.

SG. In the next PR, let's try to use ATen matmul. This will be how the integration looks like and give the best performance at this moment. Even if the error is slightly higher than the threshold, I'd accept it because (1) we know the error is localized to matmul, and (2) checking in the code allows us to investigate in parallel.